docs/developer-guide/testing.md

```diff
@@ -24,8 +25,42 @@ gobopt='--set ana.base.privatization write+lock' ./scripts/update_suite.rb
 ```
 
 ### Writing
-* Add parameters to a regression test in the first line: `// PARAM: --set warn.debug true`
-* Annotate lines inside the regression test with comments: `arr[9] = 10; // WARN`
+Regression tests use single-line comments (with `//`) as annotations.
+
+#### First line
+A comment on the first line can contain the following:
+
+| Annotation | Comment |
+| ---------- | ------- |
+| `PARAM: ` <br> (NB! space) | The following command line parameters are added to Goblint for this test. |
+| `SKIP` | The test is skipped (except when run with `./scripts/update_suite.rb group`). |
+| `NOMARSHAL` | Marshaling and unmarshaling of results is not tested on this program. |
+
+#### End of line
+Comments at the end of other lines indicate the behavior on that line:
+
+| Annotation | Expected Goblint result | Concrete semantics | Checks |
+| ---------- | ----- | ------------- | --- |
+| `SUCCESS` <br> or nothing | Assertion succeeds | Assertion always succeeds | Precision |
+| `FAIL` | Assertion fails | Assertion always fails | Precision |
+| `UNKNOWN!` | Assertion is unknown | Assertion may both <br> succeed or fail | Soundness |
+| `UNKNOWN` | Assertion is unknown | — | Intended imprecision |
+| `TODO` <br> or `SKIP` | Assertion is unknown <br> or succeeds | Assertion always succeeds | Precision improvement |
+| `NORACE` | No race warning | No data race | Precision |
+| `RACE!` | Race warning | Data race is possible | Soundness |
+| `RACE` | Race warning | — | Intended imprecision |
+| `NODEADLOCK` | No deadlock warning | No deadlock | Precision |
+| `DEADLOCK` | Deadlock warning | Deadlock is possible | Soundness |
+| `NOWARN` | No warning | — | Precision |
+| `WARN` | Some warning | — | Soundness |
+
+#### Other
+Other useful constructs are the following:
+
+| Code with annotation | Comment |
+| -------------------- | ------- |
+| `__goblint_check(1); // reachable` | Checks that the line is reachable according <br> to Goblint results (soundness). |
+| `__goblint_check(0); // NOWARN (unreachable)` | Checks that the line is unreachable (precision). |
```

Comment:
So we do not output warnings for definitely unreachable dead code in Goblint, right? And that is the reason, we can use `__goblint_check(1)` and `__goblint_check(0)` to check the reachability of a program location? Just out of interest, where is this logic of not warning in case the code is dead implemented in Goblint?

Reply:
Yes, if the code is dead, then Goblint does not output an assertion failure error for `0`, which it trivially can, if it were reached.

This logic is implemented nowhere, which is precisely why it's not easy to change. It's an implicit consequence of analysis transfer functions not being evaluated on dead code. The transfer functions not being evaluated means the `M.warn` calls never happen.